Src/IronPython/Runtime/Enumerate.cs

```diff
@@ -250,14 +250,17 @@ public class PythonEnumerator : IEnumerator {
         /// </summary>
         /// <returns>True if moving was successfull</returns>
         public bool MoveNext() {
-            if (_nextMethod == null) {
-                if (!PythonOps.TryGetBoundAttr(_baseObject, "__next__", out _nextMethod) || _nextMethod == null) {
-                    throw PythonOps.TypeError("instance has no next() method");
-                }
+
+            if (_nextMemberSite == null) {
+                _nextMemberSite = DynamicHelpers
+                    .GetPythonType(_baseObject)
+                    .GetTryGetMemberSite(DefaultContext.Default, "__next__");
             }
 
+            object nextMethod = _nextMemberSite.Target(_nextMemberSite, _baseObject, DefaultContext.Default);
```

Comment:
This is bound to be a performance hit... I don't think the behavior is clearly documented, and the test that alerted me to this behavior is really just checking for a crash.